### PR TITLE
feat: TFLint plugin manager

### LINKED ISSUE #14906: TFLint Plugins
### What would you like Renovate to be able to do?

I would like Renovate to be able to update TFLint plugin versions. TFLint plugins, by convention, are located in a `.tflint.hcl` file. TFLint plugins include, but are not limited to:

- [AWS](https://github.com/terraform-linters/tflint-ruleset-aws)
```hcl
plugin "aws" {
    enabled = true
    version = "0.13.2"
    source  = "github.com/terraform-linters/tflint-ruleset-aws"
}
```

- [Google](https://github.com/terraform-linters/tflint-ruleset-google)
```hcl
plugin "google" {
    enabled = true
    version = "0.16.1"
    source  = "github.com/terraform-linters/tflint-ruleset-google"
}
```

- [Azure](https://github.com/terraform-linters/tflint-ruleset-azurerm)
```hcl
plugin "azurerm" {
    enabled = true
    version = "0.15.0"
    source  = "github.com/terraform-linters/tflint-ruleset-azurerm"
}
```

### If you have any ideas on how this should be implemented, please tell us here.

No ideas currently. I'm not the strongest in TypeScript but I'd be willing to help implement

### Is this a feature you are interested in implementing yourself?

Maybe

### LINKED ISSUE #17880: Propose updates for tflint plugins
### What would you like Renovate to be able to do?

[TFLint](https://github.com/terraform-linters/tflint/) is a pluggable linter for [Terraform](https://terraform.io).

It can be extended by adding new plugins, which are enabled via [a configuration file](https://github.com/terraform-linters/tflint/blob/master/docs/user-guide/plugins.md) (usually, a `.tflint.hcl` file):

```hcl
plugin "foo" {
  enabled = true
  version = "0.1.0"
  source  = "github.com/org/tflint-ruleset-foo"
}
```

Plugins are fetched from GitHub Releases only at the moment.

It would be nice if Renovate detects new versions of these plugins and proposes updates on the `.tflint.hcl` file automatically.

### If you have any ideas on how this should be implemented, please tell us here.

See the [TFLint's "Configuring Plugins" page](https://github.com/terraform-linters/tflint/blob/master/docs/user-guide/plugins.md) for how the configuration file looks like.

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by multani**
## Changes

This adds a new manager to update [TFLint](https://github.com/terraform-linters/tflint/) [plugins](https://github.com/terraform-linters/tflint/blob/master/docs/user-guide/plugins.md).

There's a test repository in https://github.com/multani/test-renovate-tflint to show case the new plugin.

The code is heavily borrowed from the `terraform` manager, but is much simpler as there are much less resources to keep track of.

## Context

- Closes #17880
- closes https://github.com/renovatebot/renovate/issues/14906

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/tflint-plugin/index.ts
Comment (rarkins) [inline lib/modules/manager/tflint-plugin/index.ts:11]: Has to be in root?

File: lib/modules/manager/tflint-plugin/index.ts
Comment (rarkins) [inline lib/modules/manager/tflint-plugin/index.ts:12]: Can be removed as it's default

File: lib/modules/manager/tflint-plugin/readme.md
Comment (rarkins) [inline lib/modules/manager/tflint-plugin/readme.md:3]: Keep this all on a single markdown line

**Review by rarkins — APPROVED** [review]

**Review by rarkins — CHANGES_REQUESTED** [review]
Sorry should be request changes

File: lib/modules/manager/tflint-plugin/index.ts
Comment (multani) [reply lib/modules/manager/tflint-plugin/index.ts:11]: Not necessarily :+1: I removed the `^` anchor :+1:

File: lib/modules/manager/tflint-plugin/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/tflint-plugin/readme.md:1]:
```suggestion
Renovate maintains your [TFLint configuration file](https://github.com/terraform-linters/tflint/blob/master/docs/user-guide/config.md), and updates the plugins in the file.
```

How about this?

I don't like the non-standard verb `renovates`, we should use `updates` instead as that is a word every software engineer understands. 😉

File: lib/modules/manager/tflint-plugin/plugins.ts
Comment (HonkingGoose) [inline lib/modules/manager/tflint-plugin/plugins.ts:35]:
```suggestion
      // Therefore if we reach braceCounter == 0 then we found the end of the tflint configuration block.
```

This should be one sentence.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/modules/manager/tflint-plugin/readme.md
Comment (multani) [reply lib/modules/manager/tflint-plugin/readme.md:1]: Agree, I picked up that wording from another readme but I also found it weird :)

Comment (HonkingGoose) [reply lib/modules/manager/tflint-plugin/readme.md]: > I picked up that wording from another readme but I also found it weird :)

@multani Can you give me a link to that other readme?

**Review by HonkingGoose — DISMISSED** [review]

File: lib/modules/manager/tflint-plugin/index.ts
Comment (rarkins) [inline lib/modules/manager/tflint-plugin/index.ts:11]: Do you expect that every release MUST have a `v` prefix? Please double check that TFLint also fails if no `v` prefix is present. Otherwise we should make the `v` optional

File: lib/modules/manager/tflint-plugin/plugins.ts
Comment (rarkins) [inline lib/modules/manager/tflint-plugin/plugins.ts:65]: I would prefer if this function returned a `dep` instead of mutating it in-place

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/manager/tflint-plugin/index.ts
Comment (multani) [reply lib/modules/manager/tflint-plugin/index.ts:11]: TFlint's configuration file expects to be configured with the version without the `v` prefix, but TFLint seems to enforce it when fetching the plugins when it initializes.

For instance:

* [TFLint's function to fetch the release from GitHub](https://github.com/terraform-linters/tflint/blob/06c3ef71762ed6799490c67cdf21bf1673ff4a8d/plugin/install.go#L154-L155)
* The [`TagName` function enforces the release to start with a `v`](https://github.com/terraform-linters/tflint/blob/06c3ef71762ed6799490c67cdf21bf1673ff4a8d/plugin/install.go#L45-L50).

It's also documented in several places:

* [The `Install` function that downloads releases from GitHub](https://github.com/terraform-linters/tflint/blob/06c3ef71762ed6799490c67cdf21bf1673ff4a8d/plugin/install.go#L58-L74)
* [The developer's documentation](https://github.com/terraform-linters/tflint/blob/master/docs/developer-guide/plugins.md#4-creating-a-github-release)

I would expect releases not starting with a `v` prefix to not be discoverable by TFLint, so I would keep the the regex as it is.

Comment (rarkins) [reply lib/modules/manager/tflint-plugin/index.ts:11]: Great, thanks for looking into that.

**Review by rarkins — DISMISSED** [review]

**Comment by multani** [comment]
@rarkins I'll fix the remaining issues this morning :+1: (I stopped in the middle of the fix yesterday :( )

File: lib/modules/manager/tflint-plugin/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/tflint-plugin/readme.md:3]: Do we need to explain if the TFLint manager supports private/public repositories on `github.com`?

Comment (rarkins) [reply lib/modules/manager/tflint-plugin/readme.md:3]: Technically it can support either if the token has permissions (so same as any other manager). I'm not sure if TFlint has a need for private or not though

Comment (multani) [reply lib/modules/manager/tflint-plugin/readme.md:3]: It doesn't support plugins in private repositories at the moment (and I don't think it will ever do, AFAIU)

Comment (multani) [reply lib/modules/manager/tflint-plugin/readme.md:3]: Sounds good :+1: 

I'll let you check if you also want to add something like "this is a limitation of TFLint itself".

Comment (HonkingGoose) [reply lib/modules/manager/tflint-plugin/readme.md:3]: Yeah, let's explain _why_ we don't support private repositories. 😉 

```suggestion
Supports plugins hosted in _public_ repositories on github.com.
This is because TFLint only supports _public_ repositories.
```

**Review by HonkingGoose — DISMISSED** [review]
I'm happy with the docs now. I'll let the maintainers review the code.

File: lib/modules/manager/tflint-plugin/__fixtures__/tflint-1.hcl
Comment (viceice) [inline lib/modules/manager/tflint-plugin/__fixtures__/tflint-1.hcl:1]: please inline the smaller fixtures to reduce io

File: lib/modules/manager/tflint-plugin/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/tflint-plugin/extract.spec.ts:10]: no `?`

File: lib/modules/manager/tflint-plugin/types.ts
Comment (viceice) [inline lib/modules/manager/tflint-plugin/types.ts:12]: i don't see any usage of those manager data

**Review by viceice — COMMENTED** [review]
@zharinov should we use your parser utils instead?

**Comment by multani** [comment]
> @zharinov should we use your parser utils instead?

Do you have a HCL-like parser available?

File: lib/modules/manager/tflint-plugin/types.ts
Comment (multani) [reply lib/modules/manager/tflint-plugin/types.ts:12]: I removed it in a previous pull request but I forgot the interface here, I removed it, thanks for pointing out!

File: lib/modules/manager/tflint-plugin/extract.ts
Comment (viceice) [inline lib/modules/manager/tflint-plugin/extract.ts:53]:
```suggestion
  return deps.length ? { deps } : null;
```

Comment (multani) [reply lib/modules/manager/tflint-plugin/extract.ts]: I added a test to cover that as well

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Comment by multani** [comment]
Thanks @viceice, @HonkingGoose and @rarkins for the help and the reviews! :bow:

**Pull request merged**
